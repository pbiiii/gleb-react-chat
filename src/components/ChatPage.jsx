import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import classNames from 'classnames';
import { Typography } from '@material-ui/core/es/index';
import AppTopBar from './AppTopBar';
import ChatDrawer from './ChatDrawer';
import ChatMessageList from './ChatMessageList';
import ChatMessageInput from './ChatMessageInput/ChatMessageInput';
import ErrorMessage from './ErrorMessage';

const styles = theme => ({
    root: {
        display: 'flex',
        height: '100%',
    },
    toolbar: {
        ...theme.mixins.toolbar,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
    },
    content: {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'flex-start',
        paddingTop: '64px',
        height: '100%',
        width: '100%',
        overflow: 'hidden',
        position: 'relative',
    },
    noActiveChat: {
        alignItems: 'center',
    },
});

class ChatPage extends React.Component {
    componentDidMount() {
        const {
            fetchAllChats,
            fetchMyChats,
            match,
            socketsConnect,
            setActiveChat,
            mountChat,
        } = this.props;
        Promise.all([
            fetchAllChats(),
            fetchMyChats(),
        ]).then(() => {
            socketsConnect();
        }).then(() => {
            const { chatId } = match.params;
            if (chatId) {
                setActiveChat(chatId);
                mountChat(chatId);
            }
        });
    }

    componentWillReceiveProps(nextProps) {
        const {
            match: { params }, setActiveChat, mountChat, unMountChat,
        } = this.props;
        const { params: nextParams } = nextProps.match;

        if (nextParams.chatId && (params.chatId !== nextParams.chatId)) {
            setActiveChat(nextParams.chatId);
            unMountChat(params.chatId);
            mountChat(nextParams.chatId);
        }
    }

    render() {
        const {
            classes,
            error,
            logout,
            activeUser,
            editUser,
            chats,
            createChat,
            deleteChat,
            leaveChat,
            joinChat,
            messages,
            sendMessage,
            isConnected,
        } = this.props;
        const activeChat = chats.active;
        return (
            <div className={classes.root}>
                <CssBaseline />
                <AppTopBar
                    logout={logout}
                    activeUser={activeUser}
                    editUser={editUser}
                    activeChat={activeChat}
                    leaveChat={leaveChat}
                    deleteChat={deleteChat}
                    isConnected={isConnected}
                />
                <ChatDrawer
                    allChats={chats.all}
                    myChats={chats.my}
                    activeChat={activeChat}
                    onCreateChat={createChat}
                    isConnected={isConnected}
                />
                <main
                    className={classNames(classes.content, !activeChat && classes.noActiveChat)}
                >
                    {activeChat && (
                        <React.Fragment>
                            <ChatMessageList
                                messages={messages}
                                activeUser={activeUser}
                            />
                            <ChatMessageInput
                                sendMessage={sendMessage}
                                joinChat={() => joinChat({ chat: activeChat })}
                                activeUser={activeUser}
                                disabled={!isConnected}
                            />
                        </React.Fragment>
                    )}
                    {!activeChat && (
                        <Typography variant="h5">
                            Please select or create chat
                        </Typography>
                    )}
                </main>
                <ErrorMessage error={error} />
            </div>
        );
    }
}

export default withStyles(styles)(ChatPage);
