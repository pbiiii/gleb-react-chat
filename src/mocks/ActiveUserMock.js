import UserMock from './UserMock';

export default {
    ...UserMock,
    isMember: false,
    isCreator: true,
    isChatMember: true,
};
