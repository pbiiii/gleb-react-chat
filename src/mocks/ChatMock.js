import UserMock from './UserMock';

export default {
    _id: 'ashasdaksdmkl1mkl',
    creator: { ...UserMock, _id: 2 },
    title: 'Mock title',
};
