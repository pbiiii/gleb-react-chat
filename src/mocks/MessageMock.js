import ChatMock from './ChatMock';
import UserMock from './UserMock';

export default {
    _id: 'random_id',
    chatId: ChatMock._id,
    content: 'Mock content',
    sender: { UserMock },
    createdAt: '2019-01-10 13:00',
};
