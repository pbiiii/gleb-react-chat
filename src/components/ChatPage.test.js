/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ChatPage from './ChatPage';
import ChatMock from '../mocks/ChatMock';
import ActiveUserMock from '../mocks/ActiveUserMock';
import MessageMock from '../mocks/MessageMock';

const mockProps = {
    chats: {
        active: { ...ChatMock },
        all: [ChatMock],
        my: [],
    },
    messages: [MessageMock],
    activeUser: { ...ActiveUserMock },
    match: { params: { chatId: 'mock_new_chat_id' } },
    isConnected: true,
    socketsConnect: () => {},
    fetchAllChats: () => {},
    fetchMyChats: () => {},
    setActiveChat: () => {},
    mountChat: () => {},
    unMountChat: () => {},
    createChat: () => {},
    joinChat: () => {},
    deleteChat: () => {},
    leaveChat: () => {},
    sendMessage: () => {},
    logout: () => {},
    editUser: () => {},
};

jest.mock('./AppTopBar', () => () => 'AppTopBar');
jest.mock('./ChatDrawer', () => () => 'ChatDrawer');
jest.mock('./ChatMessageList', () => () => 'ChatMessageList');
jest.mock('./ChatMessageInput/ChatMessageInput', () => () => 'ChatMessageInput');
jest.mock('./ErrorMessage', () => () => 'ErrorMessage');

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ChatPage {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('render correctly', () => {
    const tree = renderer
        .create(<ChatPage {...mockProps} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
