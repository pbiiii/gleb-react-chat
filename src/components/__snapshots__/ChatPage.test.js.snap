// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  className="ChatPage-root-5"
>
  AppTopBar
  ChatDrawer
  <main
    className="ChatPage-content-7"
  >
    ChatMessageList
    ChatMessageInput
  </main>
  ErrorMessage
</div>
`;
